cmd/ctl/pkg/upgrade/migrate/migrate.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	fmt.Fprintln(o.Out, "Successfully migrated all cert-manager resource types. It is now safe to proceed with upgrading to cert-manager v1.7.")
+	return nil
+}
+
+func (o *Options) ensureCRDStorageVersionEquals(ctx context.Context, vers string, names []string) (bool, []*apiext.CustomResourceDefinition, error) {
+	var crds []*apiext.CustomResourceDefinition
+	for _, crdName := range names {
+		crd := &apiext.CustomResourceDefinition{}
+		if err := o.client.Get(ctx, client.ObjectKey{Name: crdName}, crd); err != nil {
+			return false, nil, err
+		}
+
+		// Discover the storage version
+		storageVersion := ""
+		for _, v := range crd.Spec.Versions {
+			if v.Storage {
+				storageVersion = v.Name
+				break
+			}
+		}
+
+		if storageVersion != vers {
+			fmt.Fprintf(o.Out, "CustomResourceDefinition object %q has storage version set to %q. You MUST upgrade to cert-manager v1.0-v1.6 before migrating resources for v1.7.\n", crdName, storageVersion)
+			return false, nil, nil
+		}
+
+		crds = append(crds, crd)
+	}
+
+	return true, crds, nil
+}
+
+func (o *Options) discoverCRDsRequiringMigration(ctx context.Context, desiredStorageVersion string, names []string) ([]*apiext.CustomResourceDefinition, error) {
+	var requireMigration []*apiext.CustomResourceDefinition
+	for _, name := range names {
+		crd := &apiext.CustomResourceDefinition{}
+		if err := o.client.Get(ctx, client.ObjectKey{Name: name}, crd); err != nil {
+			return nil, err
+		}
+		// If no versions are stored, there's nothing to migrate.
+		if len(crd.Status.StoredVersions) == 0 {
+			continue
+		}
+		// If more than one entry exists in `storedVersions` OR if the only element in there is not
+		// the desired version, perform a migration.
+		if len(crd.Status.StoredVersions) > 1 || crd.Status.StoredVersions[0] != desiredStorageVersion {
+			requireMigration = append(requireMigration, crd)
+		}
+	}
+	return requireMigration, nil
+}
+
+func (o *Options) migrateResourcesForCRD(ctx context.Context, crd *apiext.CustomResourceDefinition) error {
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: "v1",
+		Kind:    crd.Spec.Names.ListKind,
+	})
+	if err := o.client.List(ctx, list); err != nil {
+		return err
+	}
+	fmt.Fprintf(o.Out, " %d resources to migrate\n", len(list.Items))
+	for _, obj := range list.Items {
+		if err := o.client.Update(ctx, &obj); handleUpdateErr(err) != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// handleUpdateErr will absorb certain types of errors that we know can be skipped/passed on
+// during a migration of a particular object.
+func handleUpdateErr(err error) error {
+	if err == nil {
+		return nil
+	}
+	// If the resource no longer exists, don't return the error as the object no longer
+	// needs updating to the new API version.
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	// If there was a conflict, another client must have written the object already which
```

Comment:
So we know that any update to the object since we listed all objects and now would have result in the object being stored in etcd at `v1`, including i.e status update

Reply:
Yep exactly :) (because we check at the start that the storage version is set to `v1`)